waspls/exe/Main.hs

```diff
@@ -0,0 +1,74 @@
+module Main where
+
+import Data.Version (showVersion)
+import qualified Options.Applicative as O
+import qualified Paths_waspls
+import qualified Wasp.LSP.Server as LSP
+
+main :: IO ()
+main = do
+  args <- O.execParser parseArgsWithHelp
+  case command args of
+    PrintVersion -> doPrintVersion
+    Serve ->
+      if printVersion (options args)
+        then doPrintVersion
+        else LSP.run $ logFile $ options args
+  where
+    doPrintVersion = putStrLn $ showVersion Paths_waspls.version
+
+parseArgsWithHelp :: O.ParserInfo Args
```

Comment:
Yes, `execParser` does have side effects. It will print a usage message and exit if it fails to parse the options.

I'm not super happy with this `withHelp` name either, it isn't very clear, but at the same time I'm having a hard time coming up with alternatives.

Reply:
Hm I see, so this function has a lot of side effects! And it is hard to come up with a name because it is doing so many things at once.
I usually in such case start with writing a comment that explains fully what it does, and then I try to write a name that is long but fully captures what it does, and then either shorten the name or break it into multiple functions.
So let's see:

```hs
--| Parses CLI arguments. If it fails to parse them, it will print a usage message and exit.
```

So the name could be `parseArgsOrPrintUsageAndExit`. It is not very clear which condition triggers which branch of OR, but one can guess the second part happens on failure, I believe.

And I think this is it -> I can't figure out a shorter name that doesn't loose significant amount of information, and I feel like this name actually captures what is happening pretty well.

The only other solution I imagine would be to have two function: one `parseArgs`, another `printUsageAndExit`, and then at the call site we would somehow combine them, smth like `fromMaybe printUsageAndExit parseArgs` assuming parseArgs would be returning `Maybe Args` -> I am making this stuff up now I am not sure if the library we use is well suited for this, but that would look good.

Reply:
`parseArgsOrPrintUsageAndExit` sounds pretty good, although it's a little weird because that function doesn't actually do any of that, it's a description of how to do it. Probably the best name for it though.

I did take a closer look at the library. It calls this `ParserInfo` that gets passed to `execParser` a program description, but I don't think that's very clear. It does also offer more explicit ways of dealing with errors, which would look like:

```hs
main = do
  let parseResult = execParserPure defaultParserPrefs parseArgsOrPrintUsageAndExit
  args <- handleParseResult parseResult
  -- From here it's the same
```

But I think that's worse because:

- We still need the `parseArgsOrPrintUsageAndExit` variable
- Now there's `defaultParserPrefs` adding line noise
- It's exactly equivalent to the definition of `execParser`

In the docs they use a where clause to assign `opts` instead of a top-level `parseArgsOrPrintUsageAndExit`, and HLS does the same thing: https://github.com/haskell/haskell-language-server/blob/master/src/Ide/Arguments.hs#L63. Should we consider doing this, since it seems to be more standard?